_posts/2022-11-03-autoscaling-functions.md

```diff
@@ -0,0 +1,212 @@
+      com.openfaas.scale.type: capacity
+    environment:
+      write_timeout: 30s
+      exec_timeout: 30s
+      sleep_duration: 20s
+
+```
+
+We can configure autoscaling for the function via labels:
+
+* `com.openfaas.scale.min` - The minimum number of replicas to scale to.
+* `com.openfaas.scale.max` - The maximum number of replicas to scale to.
+* `com.openfaas.scale.target` - Target load per replica for scaling.
+* `com.openfaas.scale.target-proportion` - Proportion as a float of the target i.e. 1.0 = 100% of target.
+* `com.openfaas.scale.type` - Scaling mode.
+
+> All available configuration labels and there default values can be found in the [autoscaling documentation](https://docs.openfaas.com/architecture/autoscaling/)
+
+The function is running in capacity mode and the target load is 5 inflight requests (or 5 connections). The target proportion is set to 0.9.
+
+This means the autoscaler will try to maintain an average load of 90% of the target load per function by scaling the function up or down. The scaler will only scale between the configured minimum and maximum number of replicas, in this case 1 and 10.
+
+We use [K6.io](https://k6.io/) to simulate a load on the function. In this example we gradually ramp up the amount of invocation form 0 to 70 over a 4 minute period. The Grafana dashboard that we provide can be used to inspect the function and see how it performed under this load.
+
+![Grafana dashboard showing more replicas of the sleep functions are added as the amount if concurrent requests gradually increases.](/images/2022-scaling-functions/ramp-capacity.png)
+> Grafana dashboard showing more replicas of the sleep functions are added as the amount if concurrent requests gradually increases.
+
+The first graph on the top left shows the load on the function. This load is used for the scaling calculations. As you can see the load gradually increased form 0 to 70 inflight request over the course of our load test.
+
+The graph on the top right shows to number of replicas for our function. It clearly shows our function was gradually scaled up to the configured maximum amount of replicas as the load increased.
+
+The next two graphs shows the CPU and RAM usage of the function. We can see that the CPU rises as we increase the load on the function. Still it only peaks at 70mi which tells us this is quite an efficient function. The RAM usages increase as replicas of our function are added. This is an aggregated view, it shows the total amount of RAM used by the function. To get an idea of the amount of memory used by each replica you should divide it by the replica count.
+
+The bottom two graph show the invocation rate and the average invocation duration. Looking at the graph for the invocation duration we can see that each invocation takes 20 seconds to complete. This is what we would expected as the function was configured to sleep for 20 seconds.
+
+### How does this compare to the Community Edition
```

Comment:
This should probably be at the end, otherwise it may look like the headings which come after are about the community edition.

We agreed to have a section on scale to zero - for efficiency (keeping costs down when in the cloud by needing fewer nodes, and packing in more functions when on-premises) and security (if a function contains a vulnerability, it's only accessible for the time it's scaled up)